vuestorefront/models/product_template.py

```diff
@@ -0,0 +1,166 @@
+                                             store=True, readonly=True)
+    json_ld = fields.Char('JSON-LD')
+
+    @api.depends('name')
+    def _compute_website_slug(self):
+        for product in self:
+            # To make sure we don't assign ID before record is created.
+            if not isinstance(product.id, models.NewId):
+                product.website_slug = slugify_with_sfx(product.name, f"-{product.id}")
+
+    def write(self, vals):
+        res = super(ProductTemplate, self).write(vals)
+        self.env['invalidate.cache'].create_invalidate_cache(self._name, self.ids)
+        return res
+
+    def unlink(self):
+        self.env['invalidate.cache'].create_invalidate_cache(self._name, self.ids)
+        return super(ProductTemplate, self).unlink()
+
+    def _get_combination_info(
+        self,
+        combination=False,
+        product_id=False,
+        add_qty=1,
+        pricelist=False,
+        parent_combination=False,
+        only_template=False
+    ):
+        """Add discount value and percentage based."""
+        combination_info = super(ProductTemplate, self)._get_combination_info(
+            combination=combination, product_id=product_id, add_qty=add_qty, pricelist=pricelist,
+            parent_combination=parent_combination, only_template=only_template)
+
+        discount = 0
+        discount_perc = 0
+
+        if combination_info['has_discounted_price']:
+            discount = combination_info['list_price'] - combination_info['price']
+            discount_perc = combination_info['list_price'] and (
+                discount * 100 / combination_info['list_price']
+            ) or 0
+
+        combination_info.update({
+            'discount': round(discount, 2),
+            'discount_perc': int(discount_perc),
+        })
+
+        return combination_info
+
+    def get_json_ld(self):
+        self.ensure_one()
+        if self.json_ld:
+            return self.json_ld
+
+        env = self.env
+        website = env['website'].get_current_website()
+        base_url = env['ir.config_parameter'].sudo().get_param('web.base.url', '')
+        if base_url and base_url[-1:] == '/':
+            base_url = base_url[:-1]
+
+        # Get list of images
+        images = list()
+        if self.image_1920:
+            images.append('%s/web/image/product.product/%s/image' % (base_url, self.id))
+
+        json_ld = {
+            "@context": "https://schema.org/",
+            "@type": "Product",
+            "name": self.display_name,
+            "image": images,
+            "offers": {
+                "@type": "Offer",
+                "url": "%s/product/%s" % (website.domain or '', slug(self)),
+                "priceCurrency": self.currency_id.name,
+                "price": self.list_price,
+                "itemCondition": "https://schema.org/NewCondition",
+                "availability": "https://schema.org/InStock",
```

Comment:
Same on this one, we need at least TODO to update it later with actual data from the product.

Reply:
And what this `json_ld` even used for?

Reply:
From the data that is sent into it, looks like it is used for creating metadata for the products to be shown on Google Shopping option